lib/Sema/TypeCheckGeneric.cpp

```diff
@@ -136,67 +170,78 @@ OpaqueResultTypeRequest::evaluate(Evaluator &evaluator,
-  genericParamTypes.push_back(returnTypeParam);
-
   SmallVector<Requirement, 2> requirements;
-  if (constraintType->getClassOrBoundGenericClass()) {
-    requirements.push_back(Requirement(RequirementKind::Superclass,
-                                       returnTypeParam, constraintType));
-  } else {
-    auto constraints = constraintType->getExistentialLayout();
-    if (auto superclass = constraints.getSuperclass()) {
-      requirements.push_back(Requirement(RequirementKind::Superclass,
-                                         returnTypeParam, superclass));
-    }
-    for (auto protocol : constraints.getProtocols()) {
-      requirements.push_back(Requirement(RequirementKind::Conformance,
-                                         returnTypeParam, protocol));
+
+  auto opaqueReprs = collectOpaqueReturnTypeReprs(repr);
+  if (opaqueReprs.size() > 1) {
+    ctx.Diags.diagnose(repr->getLoc(), diag::more_than_one_opaque_subtype,
+                       repr);
+    return nullptr;
+  }
+
+  // TODO [OPAQUE SUPPORT]: right now we only allow one structural 'some' type,
+  // but *very* soon we will allow more than one such type.
+  for (unsigned i = 0; i < opaqueReprs.size(); ++i) {
+    auto *currentRepr = opaqueReprs[i];
+    auto *paramType = GenericTypeParamType::get(opaqueSignatureDepth, i, ctx);
+    genericParamTypes.push_back(paramType);
+
+    // Try to resolve the constraint repr in the parent decl context. It should
+    // be some kind of existential type. Pass along the error type if resolving
+    // the repr failed.
+    auto constraintType = TypeResolution::forInterface(
+                              dc, TypeResolverContext::GenericRequirement,
+                              // Unbound generics and placeholders are
+                              // meaningless in opaque types.
+                              /*unboundTyOpener*/ nullptr,
+                              /*placeholderHandler*/ nullptr)
+                              .resolveType(currentRepr->getConstraint());
+
+    if (constraintType->hasError())
+      return nullptr;
+
+    // Error out if the constraint type isn't a class or existential type.
+    if (!constraintType->getClassOrBoundGenericClass() &&
+        !constraintType->isExistentialType()) {
+      ctx.Diags.diagnose(currentRepr->getLoc(),
+                         diag::opaque_type_invalid_constraint);
+      return nullptr;
     }
-    if (auto layout = constraints.getLayoutConstraint()) {
-      requirements.push_back(Requirement(RequirementKind::Layout,
-                                         returnTypeParam, layout));
+
+    if (constraintType->hasArchetype())
+      constraintType = constraintType->mapTypeOutOfContext();
+
+    if (constraintType->getClassOrBoundGenericClass()) {
+      requirements.push_back(
+          Requirement(RequirementKind::Superclass, paramType, constraintType));
+    } else {
+      auto constraints = constraintType->getExistentialLayout();
```

Comment:
You can just add a single conformance requirement to an existential type even if it’s a composition, the GSB will expand it